### DISCUSSION TITLE
Create Tutorials that show how to use networkx with popular graph databases.

### DISCUSSION THREAD

**Discussion by tindi-plus** [networkx/networkx#6079] [Ideas]
I suggest that there should be tutorials that show how to use networkx on popular graph databases. An example is [this tutorial](https://docs.aws.amazon.com/neptune/latest/userguide/access-graph-gremlin.html) from AWS which shows how to use Gremlin on Neptune. The tutorial even includes how to use Gremlin-python via the [gremlinpython](https://pypi.org/project/gremlinpython/) package.

**Comment by MridulS** [comment]
Wouldn't this require an active instance of a graph database running somewhere all the time? We test all of our examples and tutorial regularly, so we would need to spin up a graph DB somewhere to test everything is working constantly.

We could maybe create a minimal example in the "Gallery" examples for this but I am not too excited about adding this to our tutorial.

**Reply by tindi-plus** [reply]
Hello MridulS, thank you for the response. 
What I mean is:
Question: "Can I use networkx on AWS Neptune Graph database?"
Ans: As far as I am aware, no.
Question: Can I use networkx on Neo4j?
Ans: As far as I am aware, no.
I am suggesting that for the introductory tutorial, we include examples (just mentioning names only) of graph database vendors that support networkx. Networkx is very appealing and easy to use. I was instantly thinking of where I could use it aside from datasets that can be imported into python scripts.
I am suggesting something along these lines:
"
Networkx brings graph data analysis to your fingertips. It has lots of capabilities that enable you to store, access, and manipulate graph data at scale. With networkx, creating and consuming graph data from popular graph databases such as [say, Neo4j, AWS Neptune] is easy. 
"
So it wouldn't require a database instance running. 
Taking it further, in a more detailed tutorial, we can then include:
1. steps for creating an instance of the database that supports networkx, or a link to the steps
2. Code snippets that show how to create, store and manipulate data on the mentioned database.

The person reading the tutorial who wants to test the code snippets will then create the instances by themselves.

**Reply by rossbar** [reply]
I also tend towards Mridul's opinion - NetworkX is an extremely portable/flexible library for dealing with graph data. It is of course true that you can use NetworkX in the contexts you describe, but it can also be useful in *many* other contexts and applications! NetworkX doesn't actively maintain any bridges/interfaces to cloud/externally hosted databases, so this feels like a better fit for tutorials on those providers rather than within the NetworkX project.

I agree that this might make a good tutorial or gallery example, though I too have concerns about the need to maintain an external instance for the example. Personally, the more open, community-oriented the provider, the more compelling the case would be.

**Reply by tindi-plus** [reply]
Okay. I think I understand the reasons. The database providers will be best suited to create and provide support for NetworkX on their databases.

**Comment by tindi-plus** [comment]
Thank you MridulS and Rossbar for the insightful engagement.

**Comment by rjurney** [comment]
@MridulS @rossbar I would like to add to NetworkX the ability to create nx.Graph/DiGraphs by importing a graph from the Neo4j graph database. I have created utility code that can do this in the past, but you need to do it in a transaction which can trip up new users.

As far as having a copy of the DB running all the time, that wouldn't be necessary. When unit testing, I could use Docker and maybe `docker compose` to startup [Neo4j on Docker](https://hub.docker.com/_/neo4j) and run tests against that platform. Otherwise users would specify their own connection credentials when calling the API. I haven't designed one yet, but if you think this might be a good idea then I could submit one for approval.

I don't know if this is desirable, but a neo4j adapter would be popular. I don't know if there is a modular, plug-in style architecture but if there were I am sure many graph database vendors would create one. [Memgraph](https://memgraph.com/) is one vendor in particular that has used NetworkX in their product and there are others. As a strategy, if it were possible to push some of the code into NetworkX that might limit the amount committed to internal repositories and produce more open source.

Lately I've worked on [littleballoffur](https://github.com/benedekrozemberczki/littleballoffur), [karateclub](https://github.com/benedekrozemberczki/karateclub) and now [GraphFrames](https://github.com/graphframes/graphframes) and I'd love to get into NetworkX itself.
